nifi-nar-bundles/nifi-graph-bundle/nifi-graph-processors/src/main/java/org/apache/nifi/processors/graph/ExecuteGraphQueryRecord.java

```diff
@@ -199,56 +200,70 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
                                     .getValue()))
                 );
 
-
-        boolean failed = false;
-        long delta = 0;
+        long delta;
+        FlowFile failedRecords = session.create(input);
+        WriteResult failedWriteResult = null;
         try (InputStream is = session.read(input);
              RecordReader reader = recordReaderFactory.createRecordReader(input, is, getLogger());
+             OutputStream os = session.write(failedRecords);
+             RecordSetWriter failedWriter = recordSetWriterFactory.createWriter(getLogger(), reader.getSchema(), os, input.getAttributes())
         ) {
             Record record;
-
             long start = System.currentTimeMillis();
+            failedWriter.beginRecordSet();
             while ((record = reader.nextRecord()) != null) {
                 FlowFile graph = session.create(input);
 
-                List<Map<String,Object>> graphResponses = new ArrayList<>();
-
-                Map<String, Object> dynamicPropertyMap = new HashMap<>();
-                for (String entry : dynamic.keySet()) {
-                        if(!dynamicPropertyMap.containsKey(entry)) {
+                try {
+                    Map<String, Object> dynamicPropertyMap = new HashMap<>();
+                    for (String entry : dynamic.keySet()) {
+                        if (!dynamicPropertyMap.containsKey(entry)) {
                             dynamicPropertyMap.put(entry, getRecordValue(record, dynamic.get(entry)));
                         }
-                }
+                    }
 
-                dynamicPropertyMap.putAll(input.getAttributes());
-                graphResponses.addAll(executeQuery(recordScript, dynamicPropertyMap));
+                    dynamicPropertyMap.putAll(input.getAttributes());
+                    List<Map<String, Object>> graphResponses = new ArrayList<>(executeQuery(recordScript, dynamicPropertyMap));
 
-                OutputStream graphOutputStream = session.write(graph);
-                String graphOutput = mapper.writerWithDefaultPrettyPrinter().writeValueAsString(graphResponses);
-                graphOutputStream.write(graphOutput.getBytes(StandardCharsets.UTF_8));
-                graphList.add(graph);
-                graphOutputStream.close();
+                    OutputStream graphOutputStream = session.write(graph);
+                    String graphOutput = mapper.writerWithDefaultPrettyPrinter().writeValueAsString(graphResponses);
+                    graphOutputStream.write(graphOutput.getBytes(StandardCharsets.UTF_8));
+                    graphOutputStream.close();
+                    session.transfer(graph, GRAPH);
+                } catch (Exception e) {
+                    // write failed records to a flowfile destined for the failure relationship
```

Comment:
It would be good to have a debug logger statement here so users can easily turn on logging for record failures.